## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe


<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.count()


first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [6]:
# What three item categories had the most entries?
top_three_df = df["category"].value_counts()

#Display the consumables, furniture, adn software 
top_three_df.head(3)




category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?
#use a loc
subcategory_df = df.loc[df["category"] == "consumables"]
subcategory_df["subcategory"].value_counts()



subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
client_df = df["client_id"].value_counts()
top_five_df = client_df.head(5)
top_five_df

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
#client_list = list()
client_list = list(top_five_df.index)
client_list

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?

top_client_df = df.loc[df["client_id"] == 33615]

top_client_df["qty"].sum()





64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

subtotal = df["qty"] * df["unit_price"]

df["subtotal"] = subtotal.round(2)
df.head()






first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

total_weight = df["qty"] * df["unit_weight"]
df["total_weight"] = total_weight.round(2)

#shipping_price = if df["total_weight"] <= 50: return df["total_weight"] * 10 else: return df["total_weight"] * 7

def shipping_price(total_weight):
    if total_weight <= 50:
        return total_weight * 10
    return total_weight * 7

# "Apply" the function to the amount column
df['shipping_price'] = df['total_weight'].apply(shipping_price).round(2)
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

no_tax_total_price = df["subtotal"] + df["shipping_price"]
sales_tax = no_tax_total_price * .0925
total_price = no_tax_total_price + sales_tax

df["total_price"] = total_price.round(2)
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  total_weight  shipping_price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 22 columns]

In [14]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

subtotal_cost = df["qty"] * df["unit_cost"]
total_cost = subtotal_cost + df["shipping_price"]

df["cost"] = total_cost.round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price  total_price      cost  
0            1  115164.00        787.50         5512.50    131839.08  85597.05  
1            0     523.95         31.29          312.90       914.26    629.79  
2            6     527.28         65.52          458.64      1077.12    765.18  
3            3    1056.18         35.67          356.70      1543.57   1077.35  
4            1    3902.00        928.60         6500.20     11364.40   8663.60  

[5 rows x 23 columns]

In [15]:
# Create a column for the profit of
# each line using line cost and line price

profit = df["total_price"] - df["cost"]

df["profit"] = profit.round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price  total_price      cost    profit  
0  115164.00        787.50         5512.50    131839.08  85597.05  46242.03  
1     523.95         31.29          312.90       914.26    629.79    284.47  
2     527.28         65.52          458.64      1077.12    765.18    311.94  
3    1056.18         35.67          356.70      1543.57   1077.35    466.22  
4    3902.00        928.60         6500.20     11364.40   8663.60   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
order_one_df = df.loc[df["order_id"] == 2742071]
order_one_sum_df = order_one_df["total_price"].sum().round(2)
#order_one_sum_df

order_two_df = df.loc[df["order_id"] == 2173913]
order_two_sum_df = order_two_df["total_price"].sum().round(2)
#order_two_sum_df

order_three_df = df.loc[df["order_id"] == 6128929]
order_three_sum_df = order_three_df["total_price"].sum().round(2)
#order_three_sum_df
print("Order 2742071 total: $",order_one_sum_df)
print("Order 2173913 total: $",order_two_sum_df)
print("Order 6128929 total: $",order_three_sum_df)

Order 2742071 total: $ 152811.89
Order 2173913 total: $ 162388.71
Order 6128929 total: $ 923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

#top_client_spend = df.loc[df["client_id"] == client_list]



def client_function(client_id, column):
    client_df = df.loc[df["client_id"] == client_id, column]
    return round(client_df.sum(), 2)

for client in client_list:
    print(f"{client} : ${client_function(client, 'total_price')}")



33615 : $8377308.52
66037 : $10259514.79
46820 : $9743794.36
38378 : $12906550.87
24741 : $82268892.02


In [18]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

dataframe_dict = {"client_id": client_list}

output_variable = ["qty", "shipping_price", "total_price", "cost", "profit"]

for variable in output_variable:
    dataframe_dict[variable] = [client_function(client_id, variable) for client_id in client_list]

summary_df = pd.DataFrame(dataframe_dict)
ascend_df = summary_df.sort_values("profit", ascending= False)
ascend_df

#client_id, qty, shipping_price, total_price, cost, profit

client_id     qty  shipping_price  total_price         cost       profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [62]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

renamed_df = ascend_df.rename(columns={'client_id' : 'Client ID', 'qty' : 'Quantity Sold', 'shipping_price': 'Shipping Price', 'total_price' : 'Total Price', 'cost': 'Total Cost', 'profit' : 'Total Profit'})
renamed_df

def adjust_currency(column):
    if column.name in ['Shipping Price', 'Total Price', 'Total Cost', 'Total Profit']:
        return (column / 1000000).round(2)
    else:
        return column

for column_name in renamed_df.columns:
    renamed_df[column_name]= adjust_currency(renamed_df[column_name])

# Formatting https://saturncloud.io/blog/how-to-format-numbers-in-a-python-pandas-dataframe-as-currency-in-thousands-or-millions/

renamed_df['Shipping Price'] = renamed_df['Shipping Price'].map("${}M".format)
renamed_df['Total Price'] = renamed_df['Total Price'].map("${}M".format)
renamed_df['Total Cost'] = renamed_df['Total Cost'].map("${}M".format)
renamed_df['Total Profit'] = renamed_df['Total Profit'].map("${}M".format)

renamed_df

Client ID  Quantity Sold Shipping Price Total Price Total Cost Total Profit
4      24741         239862         $5.13M     $82.27M    $45.69M      $36.58M
3      38378          73667         $3.43M     $12.91M     $9.63M       $3.27M
1      66037          43018          $1.4M     $10.26M      $7.0M       $3.26M
2      46820          75768          $1.6M      $9.74M     $7.01M       $2.74M
0      33615          64313         $1.83M      $8.38M     $6.18M        $2.2M

In [63]:
# Sort the updated data by "Total Profit" form highest to lowest
Profit_ascending = renamed_df.sort_values(['Total Profit'], ascending = False)
Profit_ascending

Client ID  Quantity Sold Shipping Price Total Price Total Cost Total Profit
4      24741         239862         $5.13M     $82.27M    $45.69M      $36.58M
3      38378          73667         $3.43M     $12.91M     $9.63M       $3.27M
1      66037          43018          $1.4M     $10.26M      $7.0M       $3.26M
2      46820          75768          $1.6M      $9.74M     $7.01M       $2.74M
0      33615          64313         $1.83M      $8.38M     $6.18M        $2.2M